conbench/entities/history.py

```diff
@@ -570,3 +588,77 @@ def _calculate_z_score(
         z_score = z_score * -1
 
     return z_score
+
+
+def _detect_shifts_with_trimmed_estimators(
+    df: pd.DataFrame, z_score_threshold=5.0
+) -> pd.DataFrame:
+    """Detect outliers and distribution shifts in historical data
+
+    Uses a z-score-based detection algorithm that uses trimmed rolling means and standard
+    deviations to avoid influence by outliers. Takes a dataframe of the same sort as
+    `_add_rolling_stats_columns_to_df`, and returning that dataframe with two additional
+    columns:
+
+    - `is_step` (bool): Is this point the start of a new segment?
+    - `is_outlier` (bool): Is this point an outlier that should be ignored?
+    """
+    tmp_df = copy.deepcopy(df)
+
+    # skip computation if no history
+    if df.shape[0] == 0:
+        tmp_df["is_step"] = pd.Series([], dtype=bool)
+        tmp_df["is_outlier"] = pd.Series([], dtype=bool)
+        return tmp_df
+
+    # pandas likes the data to be sorted
+    tmp_df.sort_values(
+        [
+            "case_id",
+            "context_id",
+            "hash",
+            "repository",
+            "timestamp",
+            "result_timestamp",
+        ],
+        inplace=True,
+        ignore_index=True,
+    )
+
+    # split / apply
+    out_group_df_list = []
+    for _, group_df in tmp_df.groupby(["case_id", "context_id", "hash", "repository"]):
+        # clean copy will only get result columns
+        out_group_df = copy.deepcopy(group_df)
+
+        group_df["mean_diff"] = group_df["mean"].diff()
+        mean_diff_clipped = copy.deepcopy(group_df.mean_diff)
+        mean_diff_clipped.loc[
+            (group_df.mean_diff < group_df.mean_diff.quantile(0.05))
+            | (group_df.mean_diff > group_df.mean_diff.quantile(0.95))
+        ] = np.nan
+        group_df["rolling_mean"] = mean_diff_clipped.rolling(
+            Config.DISTRIBUTION_COMMITS, min_periods=1
+        ).mean()
+        group_df["rolling_std"] = mean_diff_clipped.rolling(
+            Config.DISTRIBUTION_COMMITS, min_periods=1
+        ).std()
+        group_df["z_score"] = (
+            group_df.mean_diff - group_df.rolling_mean
+        ) / group_df.rolling_std
+
+        group_df["is_shift"] = group_df.z_score.abs() > z_score_threshold
+        group_df["reverts"] = group_df.is_shift & group_df.is_shift.shift(-1)
+        out_group_df["is_step"] = (
+            group_df.is_shift
+            & ~group_df.reverts
+            & ~group_df.reverts.shift(1, fill_value=False)
+        )
+        out_group_df["is_outlier"] = group_df.is_shift & group_df.reverts
+
+        out_group_df_list.append(out_group_df)
```

Comment:
Is there anything notable about this business logic that we should write down as a code comment, now that you have it at the top of your head?

Reply:
Maybe the limitations and how to tune? This won't handle some corner cases like length-2 outliers, and the z-score threshold (which is not quite the same as the previous one, because the data it operates on is filtered) can be adjusted to catch more or less.

I'll go chuck some stuff in the docstring, since the threshold is parameterized there anyway.